### ISSUE TITLE
Reused JDBC Statement Objects are not supported

### ISSUE THREAD

**Issue by eyalkoren** [elastic/apm-agent-java#221]
There is a common scenario where `PreparedStatement`s and `CallableStatement`s are created somewhere during app (or app component) initialization and the statement object reference is stored for reuse thereafter. 
We currently do not support such scenarios- when we capture statement creation we map the corresponding SQL to the created statement reference (weakly stored) and when the statement is executed - we get the SQL and then remove the mapping. 

We need to cancel this automatic mapping removal, and instead remove stale mappings through:
1. instrumentation of `java.sql.Statement#closeOnCompletion` and `java.sql.Statement#close`
2. relying on the inherent `WeakHashMap` mechanism that removes map entries of which weak key was GCd

**Comment by eyalkoren** [comment]
Actually, relying on the `close` APIs is incorrect as connection pools may rely on these APIs to return a connection to the pool (e.g. Apache DBCP), in which case the best practice would be always taking a new connection from the pool, use it, and close (i.e. return to the pool), so that would mean we have no caching of the potentially heavy connection metadata discovery

**Comment by felixbarny** [comment]
We cache the metadata per connection, typically, connections are pooled, so after some time we have cached the metadata for all connections. Additionally, we should also try to not use the data from `Connection.getMetaData()` at all.

**Comment by eyalkoren** [comment]
> We cache the metadata per connection, typically, connections are pooled, so after some time we have cached the metadata for all connections.

Right, that's why removing from cache when `close` is used is no good.

> we should also try to not use the data from Connection.getMetaData() at all.

+1 on that. Still, the metadata lookup might be expensive and we want to cache it

**Comment by felixbarny** [comment]
But this issue is just about the caching of the SQL string of `PreparedStatement`s. So on close, we can remove the statement from the cache and leave the metadata cache untouched.

**Comment by eyalkoren** [comment]
Absolutely right! My mistake...
Still, maybe relying on Statement objects becoming null can be enough.

**Comment by eyalkoren** [comment]
Changing the map to a Guava map as @felixbarny suggested.
I will not instrument `close` and `closeOnCompletion` because it seems unnecessary as long as we use weak reference for the statement object - if the app has statement leak, there is a much bigger problem that us keeping the statement String alive (we are actually probably not keeping it alive because the statement object keeps it alive itself). In addition, these concurrent maps are blocking on writes only, so in any case that statements are getting closed always, it will actually perform better if we just let the map clean the stale entries in bulks, instead of removing each time a statement is closed (a write operation).

**Comment by eyalkoren** [comment]
Importing the Guava project increased agent jar size from ~5MB to 8MB. I tried to include only the three packages of Guava directly referenced by the collection I used for PreparedStatement caching and the size decrease was not very significant (resulted in a 6.7MB jar). Considering that we may be more susceptible to future changes in the internal Guava dependencies, we decided to just import the entire module.
The dependency is now defined in the agent-core module, which means it can be used wherever needed in the agent code. In addition, the concurrency level is set to 16, overriding the default level of 4.
